source/common/quic/envoy_quic_stream.cc

```diff
@@ -102,10 +102,85 @@ void EnvoyQuicStream::encodeTrailersImpl(spdy::Http2HeaderBlock&& trailers) {
   onLocalEndStream();
 }
 
-void EnvoyQuicStream::encodeMetadata(const Http::MetadataMapVector& /*metadata_map_vector*/) {
-  // Metadata Frame is not supported in QUICHE.
-  ENVOY_STREAM_LOG(debug, "METADATA is not supported in Http3.", *this);
-  stats_.metadata_not_supported_error_.inc();
+std::unique_ptr<Http::MetadataMap>
+EnvoyQuicStream::metadataMapFromHeaderList(const quic::QuicHeaderList& header_list) {
+  auto metadata_map = std::make_unique<Http::MetadataMap>();
+  for (const auto& [key, value] : header_list) {
+    (*metadata_map)[key] = value;
+  }
+  return metadata_map;
+}
+
+namespace {
+
+// Returns a new `unique_ptr<char[]>` containing the characters copied from `str`.
+std::unique_ptr<char[]> dataFromString(const std::string& str) {
+  auto data = std::make_unique<char[]>(str.length());
+  memcpy(&data[0], str.data(), str.length()); // NOLINT(safe-memcpy)
+  return data;
+}
+
+void serializeMetadata(const Http::MetadataMapPtr& metadata, quic::QuicStreamId id,
+                       absl::InlinedVector<quiche::QuicheMemSlice, 2>& slices) {
+  quic::NoopDecoderStreamErrorDelegate decoder_stream_error_delegate_;
+  quic::QpackEncoder qpack_encoder_(&decoder_stream_error_delegate_,
+                                    quic::HuffmanEncoding::kDisabled);
+
+  spdy::Http2HeaderBlock header_block;
+  for (const auto& [key, value] : *metadata) {
+    header_block.AppendValueOrAddHeader(key, value);
+  }
+
+  // The METADATA frame consist of a frame header, which includes payload
+  // length, and a payload, which is the QPACK-encoded metadata block. In order
+  // to generate the frame header, the payload needs to be generated first.
+  std::string metadata_frame_payload =
+      qpack_encoder_.EncodeHeaderList(id, header_block,
+                                      /* encoder_stream_sent_byte_count = */ nullptr);
+  std::string metadata_frame_header =
+      quic::HttpEncoder::SerializeMetadataFrameHeader(metadata_frame_payload.size());
+
+  slices.emplace_back(dataFromString(metadata_frame_header), metadata_frame_header.length());
+  slices.emplace_back(dataFromString(metadata_frame_payload), metadata_frame_payload.length());
+}
+
+} // namespace
+
+void EnvoyQuicStream::encodeMetadata(const Http::MetadataMapVector& metadata_map_vector) {
+  if (!http3_options_.allow_metadata()) {
+    // Metadata Frame is not supported in QUICHE.
+    ENVOY_STREAM_LOG(debug, "METADATA is not supported in Http3.", *this);
+    stats_.metadata_not_supported_error_.inc();
+    return;
+  }
+  if (quic_stream_.write_side_closed()) {
+    IS_ENVOY_BUG("encodeMetadata is called on write-closed stream.");
+    return;
+  }
+  ASSERT(!local_end_stream_);
```

Comment:
Why is this ASSERT held?

Reply:
Hm. Mostly because we have the same ASSERT in encodeHeaders and friends. I assumed it was valid because if `local_send_stream_` were true, then so would `write_side_closed()` and we would early return from this method on line 158. WDYT?